pupil_src/shared_modules/video_capture/file_backend.py

```diff
@@ -307,64 +318,59 @@ def idx_to_pts(self, idx):
-            elif target_entry.container_idx != self.current_container_index:
-                self.setup_video(target_entry.container_idx)
-            try:
-                av_frame = next(self.next_frame)
-                index = self._convert_frame_index(av_frame.pts)
-                if index == self.target_frame_idx:
-                    break
-                elif index < self.target_frame_idx:
-                    pass
-            except StopIteration as stop:
-                raise EndofVideoError from stop
-
-            # if self.loop:
-            #     logger.info("Looping enabled. Seeking to beginning.")
-            #     self.seek_to_frame(0)
-            #     return self.get_frame()
-            # else:
-            #     logger.debug(
-            #         "End of videofile %s %s"
-            #         % (self.current_frame_idx, len(self.timestamps))
-            #     )
-            #     raise EndofVideoError("Reached end of video file")
-
+        except IndexError:
+            raise EndofVideoError
+        if target_entry.container_idx == -1:
+            return self._get_fake_frame_and_advance(target_entry.timestamp)
+        elif target_entry.container_idx != self.current_container_index:
+            self.setup_video(target_entry.container_idx)
+        try:
+            av_frame = next(self.next_frame)
+            if not av_frame:
+                raise EndofVideoError
+            index = self._convert_frame_index(av_frame.pts)
+        except StopIteration as stop:
+            return self._get_fake_frame_and_advance(target_entry.timestamp)
+            raise EndofVideoError from stop
         self.target_frame_idx = index + 1
         self.current_frame_idx = index
         return Frame(target_entry.timestamp, av_frame, index=index)
 
+    def _get_fake_frame_and_advance(self, ts):
+        self.current_frame_idx = self.target_frame_idx
+        self.target_frame_idx += 1
+        return FakeFrame(ts, self.target_frame_idx)
+
     @ensure_initialisation(fallback_func=lambda evt: sleep(0.05))
     def recent_events_external_timing(self, events):
         try:
             last_index = self._recent_frame.index
         except AttributeError:
             # Get a frame at beginnning
-            frame = self.get_frame()
-            self._recent_frame = frame
             last_index = -1
         # Seek Frame
+        frame = None
         pbt = self.g_pool.seek_control.current_playback_time
         ts_idx = self.g_pool.seek_control.ts_idx_from_playback_time(pbt)
-        if ts_idx < last_index or ts_idx > last_index + 1:
+        if ts_idx == last_index:
+            frame = self._recent_frame.copy()
+        elif ts_idx < last_index or ts_idx > last_index + 1:
             self.seek_to_frame(ts_idx)
 
         # Normal Case to get next frame
-        if not self.play:
+        try:
+            frame = frame or self.get_frame()
+        except EndofVideoError:
+            if self.loop:
```

Comment:
`loop` does not need to be supported in `recent_events_external_timing`. I should have made this more clear before.

Reply:
Oh, I missed understand, so we can delete all the loop part, right?

Reply:
> OPD should not loop.
But we still have to keep this `loop` code in `recent_events_own_timing`, right?

Reply:
Correct, that is required for File_Source usage in Capture.